### DISCUSSION TITLE
minimap in the corner & chance to hit + damage dealt vanilla vs devX

### DISCUSSION THREAD

**Discussion by JohnnyR8** [diasurgical/DevilutionX#7433] [General]
thanks guys for this phenomenal mod/port... really like going back to old days :-)

i have 2 question:

on the main devilutionX page on the picture is a minimap in the right corner.. how can i get this in win build?
[https://github.com/diasurgical/devilutionX](https://github.com/diasurgical/devilutionX)

this minimap is also seeable in the mac build:
[https://github.com/diasurgical/devilutionX/issues/6936](https://github.com/diasurgical/devilutionX/issues/6936)

the other question has to do with chance to hit & dealt damage:

i know that devX creators changed the displayed cth: 
[https://www.reddit.com/r/Diablo1/comments/zcxuwb/did_devilutionx_up_the_base_hit_chance/?rdt=60706](https://www.reddit.com/r/Diablo1/comments/zcxuwb/did_devilutionx_up_the_base_hit_chance/?rdt=60706)

in vanilla GOG - Diablo 1.09 the starting rogue has armorclass: 6 and cth: 65%.

in DevX 1.5.3 -  armorclass 8 and cth: 101%. 

that big gap & numbers wouldn't bother me if it plays identical but i tested something out with the Butcher:

rogue savegame:
[single_1.zip](https://github.com/user-attachments/files/17134468/single_1.zip)

when you simply open the door and click the butcher dead in vanilla in most of the cases you have left over 1 big Healthpotion.

with 1.5.3 devX (armorclass: 33, cth: 112%) it's impossible to kill him using all healthpotions (one-third to ~half of the butchers health bar is left, when you die) ... from that i conclude something from the dealt damage in devX is different than in vanilla Diablo (armorclass: 19, cth: 67%)...

i did this test with all devX windows versions and the last version which i found identical to kill the butcher to vanilla is 1.3.0 (armorclass: 19, cth: 112%).

it has nothing to do with the displayed armor class, because devX 1.4.0 (armorclass: 19, cth: 112) acts exactly the same as devX 1.5.3 (armorclass: 33, cth: 112%).

this means devX 1.4.0 is the first version, where it takes much longer to kill the bastard! ;-)

i also tested if it had to do with the framefrate of the fast slaying butcher by limiting devX to 20 fps like the vanilla game, but this makes no difference. [https://github.com/diasurgical/devilutionX/discussions/6971](https://github.com/diasurgical/devilutionX/discussions/6971) 

hopefully someone can bring up some light what's happening here... i really don't want to bash this great mod/port... hopefully it's only a little bug or something wrong on my side :-)

EDIT:
tested it:

in vanilla D the player gets hit ~ 11 times before the butcher falls, 
in 1.3.0 the player gets hit ~ 14 times before the butcher gets down,  (very comparable to vanilla D) 
and with 1.5.3 the butcher hits you ~ 25 times before the player dies

**Comment by qndel** [comment]
we haven't changed any formulas

**Comment by julealgon** [comment]
@JohnnyR8 the numbers you see in vanilla 1.09 are just incorrect. The ones that DevilutionX displays are the final, correct numbers.

**Comment by JohnnyR8** [comment]
thx for your reply guys... i know... you can simple test it out with the savegame as described

**Reply by AJenbo** [reply]
If you can build the game then you could use git bisect to narrow it down to the exact change between 1.3.0 and 1.4.0

**Reply by StephenCWills** [reply]
The player sure seems to get hit a lot more in 1.4.0 than 1.3.0. That's perfectly normal because combat uses random numbers. To put your mind at ease, I went ahead and confirmed that we didn't break the computations for CTH. If I were to guess, the discrepancy is most likely because of a change to the RNG sequence used for combat due to #3744.

**Reply by JohnnyR8** [reply]
yes i tested it: 

in vanilla D the player gets hit ~ 11 times before the butcher falls, in 1.3.0 the player gets hit ~ 14 times before the butcher gets down, and with 1.5.3 the butcher hits you ~ 25 times before the player dies

**Reply by StephenCWills** [reply]
> The player sure seems to get hit a lot more in 1.4.0 than 1.3.0.

I guess what I was trying to do was make a qualitative statement about how long the player stays in stunlock between the two versions. I didn't make that very clear. What should have been clear, though, is that this is all normal because the game uses RNG. Your tests don't indicate that anything is wrong.

**Reply by DakkJaniels** [reply]
> yes i tested it: 
> 
> in vanilla D the player gets hit ~ 11 times before the butcher falls, in 1.3.0 the player gets hit ~ 14 times before the butcher gets down, and with 1.5.3 the butcher hits you ~ 25 times before the player dies

How many games did you run in each version to test this? At least a few hundred, I hope.

**Reply by JohnnyR8** [reply]
sry guys, but some reactions sounds like from a caught child... 

i have made clear in my initial post how much i admire the dedication of this mod/port, don't want to take away anything from it or bash it and be thankful that people keep doing this in their free time!

but its evident, that something is working different than in vanilla D - i haven't used the word "wrong" - we are talking about a game, not real life... simply different

it's easy to reproduce in (~ 30 seconds a run) for everybody with the included savegame and at least after 10 tries anyone should be able to see the clear difference...

**Reply by StephenCWills** [reply]
What are you talking about? I tried it too. Yes, the outcome was different. Because of the random numbers.

**Reply by JohnnyR8** [reply]
"At least a few hundred (tries), I hope"    ;-)

if the random numbers difference would be as close as with 1.3.0 i barely wouldn't have noticed it... but from 1.4.0 upwards its impossible to kill the butcher... 
and i its likely that this behavior goes to the whole game... the butcher is simply the first enemy which has enough health, to test it out...

i thought we are talking about a project which hasn't changed numbers and therefore the gameplay should be identical to the Vanilla Diablo experience?

that all is positive criticism and where should i post it, if not in this place?

or do you don't want a close as possible vanilla experience?

**Reply by StephenCWills** [reply]
He's saying that a few hundred tries would be necessary to prove statistically that something has changed, given that the outcome is random.. due to the random numbers. And keep in mind that when you `Load Game`, you get the same sequence of random numbers each time. So the level of effort you need to prove anything using this approach is nontrivial.

And I am not saying that nothing has changed. I explicitly said that the sequence of random numbers has changed. But you are heavily implying that we changed combat, and we have repeatedly told you that we haven't. I don't know what else you expect us to do here.

**Reply by JohnnyR8** [reply]
for the player it‘s irrelevant what has changed under the hood… the outcome of the battle should be identical to vanilla experience

i am leaving it for here and hope get‘s fixed in upcoming releases :-)

**Reply by AJenbo** [reply]
I don't think you understand. Statistically the battles are identical, the only change is which are randomly harder and which are randomly easier. The change you are seeing is very specific to the save game you have, if you start a new one it could be just as easy as this one was in 1.3.0, and if you start a new one in 1.3.0 it could be just as hard as this one is in 1.4.0.
There is nothing to fix here, you where just less lucky with this save, try starting a new game and you might get lucky again.

**Reply by JohnnyR8** [reply]
thx for this explaining comment Ajenbo :-)

this leads to some more detailed questions:

where random numbers also present in vanilla D?

or where the added in 1.4.0, or maybe earlier because 1.3.0 also doesn't behave completly ident (but very compareable to vanilla) with my savegame and why?

is there a possibilty to turn this feature off without coding? i simply want your great mod as near as possible to vanilla D 


and one complete other question: rogue cth 101%, armorclass 8: 
a starting value over 100% irritates a little... vanilla has cth 65%, armorclass 6... is that as intended?

**Reply by DakkJaniels** [reply]
The game is literally based on random number generation, you can't turn it off.

Regarding the display, >100% chance to hit does not mean you always hit. That is your starting chance, before reducing it based on monster AC and level. There is no difference from how vanilla calculated this, DevilutionX just displays the bonuses that vanilla did not include in the displayed number.

**Reply by AJenbo** [reply]
> is there a possibilty to turn this feature off without coding?

The difference is imply that the game roles the dice a few more times before the fight with butcher starts. Lets say that in your save game the next random numbers are 1, 3, 4, 6

DevilutionX 1.3.0 using random numbers:
1 <--- build dungeon
3 <--- place monster in a wall
4 <--- start fight with Butcher
6 <--- start fight with fallen one

DevilutionX 1.4.0 using random numbers:
1 <--- build dungeon
3 <--- place monster in a wall
4 <--- move the monster out of wall (new bugfix in 1.4.0)
6 <--- start fight with Butcher

Because of the new fix the random number given to the butcher is now 6 instead of 4 in your save game, so for that game he turned out harder. But if you start a new game and the random number in that game is 5, 2, 6, 2 butcher should be easier there since he gets a 2 instead of a 6.

**Reply by AJenbo** [reply]
>  a starting value over 100% irritates a little... vanilla has cth 65%, armorclass 6... is that as intended?

A lot of the information in the panel was not actually showing the real numbers as they forgot to update the panel when adjusting the game. We have corrected this so that you now see the real numbers, but we have not changed the actual numbers in the game. It's only visual so that you can better know what is really happening.

**Reply by StephenCWills** [reply]
I take it back. There actually is something wrong with this save file. The following screenshots were taken from a debug session in Devilution.

---

Here is the `monster` array after dungeon generation. The Butcher's data is located in `monster[4]`. He has 50 hit, which matches up with the monster data table and Jarulf's Guide.

![image](https://github.com/user-attachments/assets/72a80ee1-486a-46d6-9be9-e9769f66a6b7)

---

Here is what happens after we load all the monster data out of the save file. The Butcher is now located in both `monster[4]` and `monster[11]`, and the latter has 25 hit. The monster at 4 is inactive so the one you're fighting is `monster[11]`.

![image](https://github.com/user-attachments/assets/6c53c4ee-1eca-475f-b6c9-cdd2fcb74134)

---

This doesn't really explain what's different between 1.3.0 and 1.4.0, but the Butcher's base hit chance in vanilla is halved. That what makes him so much easier.

One additional thing I noticed. When debugging in DevilutionX, dungeon generation places The Butcher in `Monsters[11]`. That might explain why he's in that location in the save file, but it doesn't really explain where the 25 hit came from.

**Reply by StephenCWills** [reply]
Ah, actually, I just thought of what could have happened. The reason The Butcher is in `Monsters[11]` is because objects got placed first, meaning all the Skeletons from barrels and sarcophagi would get initialized before The Butcher.

The code in `loadsave.cpp` claims the monster's `toHit` is already initialized.

https://github.com/diasurgical/devilutionX/blob/b78ec233625201a57c467e4db73ff3ce3a50637c/Source/loadsave.cpp#L681

However, it would have been initialized in `InitMonster()` during dungeon generation. And the monster located in `Monsters[11]` depends on whether you are playing a version before 1.4.0 or after. So if you take a save from 1.4.0 back into 1.3.0, you'd end up with a Butcher that has the wrong base hit chance.

**Reply by AJenbo** [reply]
Hm this would also affect someone moving back in to vanilla. I guess the conclusion is that 1.4.0 behaves correctly but you can't use a save for comparison if it has been written to by 1.4/1.5

**Reply by JohnnyR8** [reply]
thank you all for your detailed explanations - i really appreciate it... keep your dedication for this great alltime classic :-)

as additional info: the savegame was created with release 1.5.3

**Reply by StephenCWills** [reply]
> I guess the conclusion is that 1.4.0 behaves correctly but you can't use a save for comparison if it has been written to by 1.4/1.5

It's a little more complicated than that. You can take a save from 1.4.0 back to vanilla, and it will work fine because vanilla reads the CTH value that 1.4.0 writes to the file. The issue here is when you take the save from 1.4.0 to 1.3.0. 1.3.0 applies the wrong CTH value to the monster and puts it back in the save file. Now both 1.3.0 and vanilla are wrong.

**Comment by AJenbo** [comment]
> on the main devilutionX page on the picture is a minimap in the right corner.. how can i get this in win build?

Install 1.6.0 once it is released ... or download the test builds

**Comment by deltaoak** [comment]
> on the main devilutionX page on the picture is a minimap in the right corner.. how can i get this in win build?

check the keybind for "Cycle Map Type"
